pkg/controller/argocd/hooks.go

```diff
@@ -0,0 +1,33 @@
+package argocd
+
+import (
+	"sync"
+
+	argoprojv1alpha1 "github.com/argoproj-labs/argocd-operator/pkg/apis/argoproj/v1alpha1"
+)
+
+var (
+	mutex sync.RWMutex
+	hooks = []Hook{}
+)
+
+// Hook changes resources as they are created or updated by the
+// reconciler.
+type Hook func(*argoprojv1alpha1.ArgoCD, interface{}) error
+
+// Register adds a modifier for updating resources during reconciliation.
+func Register(h Hook) {
+	mutex.Lock()
+	defer mutex.Unlock()
+	hooks = append(hooks, h)
+}
+
+func applyReconcilerHook(cr *argoprojv1alpha1.ArgoCD, i interface{}) error {
+	for _, v := range hooks {
```

Comment:
I appreciate that it's unlikely to happen in real-world use, but it's best-practice to lock and unlock the mutex in here.